builder/azure/arm/step_deploy_template.go

```diff
@@ -298,21 +286,48 @@ func (s *StepDeployTemplate) deleteDeploymentResources(ctx context.Context, subs
 	}
 
 	resources := map[string]string{}
-
+	var vmID *virtualmachines.VirtualMachineId
+	var nicID *commonids.NetworkInterfaceId
 	for _, deploymentOperation := range deploymentOperations.Items {
 		// Sometimes an empty operation is added to the list by Azure
 		if deploymentOperation.Properties.TargetResource == nil {
 			continue
 		}
-
 		resourceName := *deploymentOperation.Properties.TargetResource.ResourceName
 		resourceType := *deploymentOperation.Properties.TargetResource.ResourceType
 
-		s.say(fmt.Sprintf("Adding to deletion queue -> %s : '%s'", resourceType, resourceName))
-		resources[resourceType] = resourceName
-
+		// Create vm and NIC ID to use them to delete later
+		switch resourceType {
+		case "Microsoft.Compute/virtualMachines":
+			vmIDDeref := virtualmachines.NewVirtualMachineID(subscriptionId, resourceGroupName, resourceName)
+			vmID = &vmIDDeref
+		case "Microsoft.Network/networkInterfaces":
+			nicIDDeref := commonids.NewNetworkInterfaceID(subscriptionId, resourceGroupName, resourceName)
+			nicID = &nicIDDeref
+		default:
+			s.say(fmt.Sprintf("Adding to deletion queue -> %s : '%s'", resourceType, resourceName))
+			resources[resourceType] = resourceName
+		}
 	}
 
+	// Always delete VM first, then delete NIC, this way we avoid conflicts on deletion
+	if vmID != nil {
+		s.say(fmt.Sprintf("Attempting deletion -> %s : '%s'", "Microsoft.Compute/virtualMachines", vmID.VirtualMachineName))
+		if err := s.client.VirtualMachinesClient.DeleteThenPoll(pollingContext, *vmID, virtualmachines.DefaultDeleteOperationOptions()); err != nil {
```

Comment:
To be clear, does `DeleteThenPoll` gets retried if moved out of the loop below? I can see a world where this fails for many reasons that may be transient (temporary connection loss, busy servers, endpoint failure for whatever reason, etc.) , failing completely feels suboptimal if so.

Reply:
It does not, it may be worth adding them to their own retry blocks, but separate from each other, ideally we don't leave behind VMs which I believe is why everything was in a retry block, however this lead to a lot of confusion and frustration from users who think the errors are a bit annoying.

Do you think it'd be worthwhile to create two extra retry blocks, one for VMs and one for Network interfaces, to ensure they both get deleted, even if the API flakes (tbh I've never seen this happen)